Distributed nearest callback

aprokop opened the pull request:
[The 3-round rework of #737 will be pushed in that branch once this PR has been merged].

Works for both APIv1 and APIv2 (that was quite tricky).

Introduces two ways to do a nearest callback in the distributed setting:
- 2-round communication for a constrained callback

Constrained meaning that
- It is allowed to be called on not final matches
- For each match, it outputs exactly one result

Constrained callback is indicated with a new tag.

Current limitations:
- I have not thought yet about post-callbacks and whether they should be allowed
  To be fair, even with the spatial callbacks I'm not sure what that means

aprokop commented on src/details/ArborX_Callbacks.hpp:
```diff
@@ -44,6 +46,25 @@ struct DefaultCallback
   }
 };
 
+#ifdef ARBORX_ENABLE_MPI
+struct ConstrainedNearestCallbackTag
+{};
+
+struct DefaultCallbackWithRank
+{
+  using tag = ConstrainedNearestCallbackTag;
+
```
@dalg24 I know you don't like this approach (similar to the discussion we had in #675). What alternative do you propose?

dalg24 replied:
You define a decorator that wraps the callback and you check whether the callback is decorated.

Something like
```C++
template <class Callback>
struct DecoratedCallback {
  Callback callback_;
  template <class... Args>
  void operator()(Args&&... args) const { callback_((Args&&)args...); }
};

template <class Callback>
struct is_decorated_callback : std::false_type {};

template <class Callback>
struct is_decorated_callback<DecorateCallback<Callback>> : std::true_type {};
```
Try to come up with more descriptive names.
I am not convinced this belongs to the general callback header, you might want to move it to the distributed headers.

aprokop replied:
Opted for the decorator instead of updating `operator()`.

aprokop commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,212 @@
+namespace Details
+{
+
+template <typename Tree, typename Callback, typename OutValue, bool UseValues>
```
The `OutValue` argument is really annoying. Currently, no avoiding it. The issue is that we can't determine from the callback's signature what the output value type is going to be. The way we are doing things right now is that determine that from the storage arrays passed in by a user.

aprokop commented on src/details/ArborX_DetailsLegacy.hpp:
```diff
@@ -84,7 +84,7 @@ struct LegacyCallbackWrapper
                                   Output const &out) const
   {
-    _callback(predicate, value.index, out);
+    _callback(predicate, (int)value.index, out);
```
This is a more precise treatment for APIv1. I also needed this due to the default distributed callback now being templated on the value, resulting in the `unsigned -> signed` loss of precision errors.

aprokop replied:
Added a comment.

dalg24 replied:
Needed for aggregate initialization

dalg24 commented on test/tstDistributedTree.cpp:
```diff
@@ -31,6 +31,41 @@ namespace tt = boost::test_tools;
+  KOKKOS_FUNCTION void operator()(Predicate const &, Value const &value,
+                                  OutputFunctor const &out) const
+  {
+    out({rank, value});
```
Why do we need both this one and `DefaultCallbackWithRank`?
One aggregates the rank first and the value second, the other swaps the two.

aprokop replied:
We don't *need* it. Rather, this is just a custom callback that does this.

dalg24 replied:
Maybe add a comment to that effect then.

aprokop replied:
Done.

aprokop commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+auto declare_callback_constrained(Callback const &callback)
+{
+  return ConstrainedDistributedNearestCallback<Callback>{callback};
+}
+
+} // namespace Experimental
+
+namespace Details
+{
+
+struct DefaultCallbackWithRank
+{
+  int _rank;
+
+  template <typename Predicate, typename Value, typename OutputFunctor>
+  KOKKOS_FUNCTION void operator()(Predicate const &, Value const &value,
+                                  OutputFunctor const &out) const
+  {
+    out({value, _rank});
+  }
+};
+
+template <class Callback>
+struct is_constrained_callback : std::false_type
+{};
+template <class Callback>
+struct is_constrained_callback<
+    Experimental::ConstrainedDistributedNearestCallback<Callback>>
+    : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallback> : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallbackWithRank> : std::true_type
+{};
+
+template <class Callback>
+inline constexpr bool is_constrained_callback_v =
+    is_constrained_callback<Callback>::value;
```
@dalg24 Please look at this approach.

dalg24 replied:
ok

aprokop commented on src/details/ArborX_DetailsLegacy.hpp:
```diff
@@ -96,24 +96,10 @@ struct LegacyDefaultCallback
 };
 
-#ifdef ARBORX_ENABLE_MPI
-struct LegacyDefaultCallbackWithRank
```
I moved this to distributed headers and renamed. It no longer is being called only from APIv1, but also from within APIv2.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+    Experimental::ConstrainedDistributedNearestCallback<Callback>>
+    : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallback> : std::true_type
+{};
```
It is non-obvious to me why we need a specialization for the default callback (haven't thought too hard about it) so it might be worth a comment.

aprokop replied:
We need it because we use default callback internally when implementing a distributed nearest query without user provided callback.

aprokop replied:
Added a comment.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+      min_corner[d] -= distance;
+      max_corner[d] += distance;
+    }
+    return intersects(box);
```
This enlarging the box can yield false positive right?  Is that potentially an issue?
We might want to add a comment.

aprokop replied:
The behavior is unchanged since `master`. It can do false positive for the upper tree, resulting in more work to be done. From a user perspective, there is no false positives.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto const ray = getGeometry(x.predicates(i));
+    return intersects(ray);
```
Remind me, don;t we only intersect with geometries in front of the origin?
In that case should we be moving the origin backwards for the predicate?

aprokop replied:
The behavior is unchanged from the `master`. Ray is special since the meaning of the nearest is different for it (nearest only among those intersected). Since the intersection of rank boxes would provide the correct ordering, this does not require a change.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+    if constexpr (UseValues)
+    {
+      OutValue out_value;
+      [[maybe_unused]] int count = 0;
```
Why did you add the `[[maybe_unused]]` attribute here?
It is being incremented below.  Were you actually seeing warning with `NDEBUG` defined?

aprokop replied:
The compiler [complained](https://cloud.cees.ornl.gov/jenkins-ci/job/ArborX/job/PR-1075/2/gcc/folder.1897384215/) about it.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+      [[maybe_unused]] int count = 0;
+      _callback(query, value, [&](OutValue const &ov) {
+        out_value = ov;
+        ++count;
```
Doesn't this incrementing the counter need to be atomic?

aprokop replied:
This assumes that each query is executed by a single thread. If that assumption is wrong, we don't have an easy way to check this, unless we create a full view and do atomics.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+        out_value = ov;
+        ++count;
+      });
+      KOKKOS_ASSERT(count == 1);
```
Why can it only be called once?  It might be worth a comment.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+      : _tree(tree)
+      , _callback(callback)
+  {
+    // NOTE cannot have extended __host__ __device__  lambda in constructor with
+    // NVCC
```
```suggestion
    // NOTE cannot have extended __host__ __device__ lambda in constructor with
    // NVCC
```

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,271 @@
+            Kokkos::WithoutInitializing,
+            "ArborX::DistributedTree::query::nearest::reverse_permutation"),
+        n);
+    if (!_tree.empty())
```
I would prefer `if(tree.empty()) return;` first thing in this function

aprokop commented on src/ArborX_DistributedTree.hpp:
```diff
@@ -190,39 +190,14 @@ class DistributedTree<MemorySpace, Details::LegacyDefaultTemplateValue,
-          });
-    }
+    int comm_rank = -1;
+    if (base_type::getComm() != MPI_COMM_NULL)
```
Not for the default constructed one.

aprokop commented:
@dalg24 Should have addressed all your comments with one exception.

Regarding atomics, I think it is fine as it is. I did not do atomics, as I think it is fine as it is. The only way I can see us triggering it if we call callbacks on the on-node nearest neighbor search results in parallel. We don't currently do that, nor do we have any plans to do it, as nearest neighbors require priority trees. The only exception to the latter is k = 1 specialization in BruteForce (not currently implemented), but even that would result in a single answer and thus no parallel callback calls for the same query.

dalg24 commented on src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:
```diff
@@ -0,0 +1,285 @@
+{
+
+// Constrained callback is a callback that a user promises to:
+// - be not pure
```
I don't get the "not pure".  Plz explain.

dalg24 replied:
Please clarify that takes two arguments

dalg24 reviewed (commented):
Just a few last comments

dalg24 reviewed (approved):
Merge after you add the comments I asked for

aprokop commented:
4 builds (all CUDA and SYCL) timed out, though 2 of them made through the first make+ctest. 3 builds passed. Given that there's been only comment change since 94330e2, I'm comfortable merging this.
